File: prototype-pattern/python/src/Shape.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""

import copy


class Shape():
    width = 0
    height = 0
    color = ''
    type = None

    def __init__(self):
      print("Shape init")

    def draw(self):
        print('Shape::draw()')

    def set_width(self, width):
        self.width = width

    def get_width(self):
        return self.width

    def get_height(self):
        return self.height

    def set_height(self, height):
        self.height = height

    def set_color(self, color):
        self.color = color

    def get_color(self):
        return self.color

    def clone(self):
        return copy.copy(self)
        # return copy.deepcopy(self)

    def to_string(self):
      # return self.__dict__
      return '{{ width = {0}, height = {1}, type = {2}, color = {3} }}'.format(
            self.width, self.height, self.type, self.color)

File: prototype-pattern/java/src/Shape.java
package src;

// 基础抽象类
public abstract class Shape implements Cloneable {

  private int width;
  private int height;
  private String color = "";
  protected String type;

  public Shape() {

  }

  public String getType() {
    return type;
  }

  // 抽象方法，子类覆盖
  public abstract void draw();

  public void setWidth(int width) {
    this.width = width;
  }

  public int getWidth() {
    return this.width;
  }

  public int getHeight() {
    return this.height;
  }

  public void setHeight(int height) {
    this.height = height;
  }

  public void setColor(String color) {
    this.color = color;
  }

  public String getColor() {
    return this.color;
  }

  // 克隆方法
  public Object clone() {
    Object clone = null;
    try {
      clone = super.clone();
    } catch (CloneNotSupportedException e) {
      e.printStackTrace();
    }
    return clone;
  }

  @Override
  public String toString() {
    return String.format("{width = %s, height = %s, type = %s, color = %s }",
        this.width, this.height, this.type, this.color);
  }
}
